src/blueprint/into_schema.rs

```diff
@@ -38,32 +40,44 @@ fn to_type(def: &Definition) -> dynamic::Type {
-                    let req_ctx = ctx.ctx.data::<Arc<RequestContext>>().unwrap();
-                    let field_name = &field.name;
-                    match &field.resolver {
-                        None => {
-                            let ctx = EvaluationContext::new(req_ctx, &ctx);
-                            FieldFuture::from_value(
-                                ctx.path_value(&[field_name]).map(|a| a.to_owned()),
-                            )
-                        }
-                        Some(expr) => {
-                            let expr = expr.to_owned();
-                            FieldFuture::new(async move {
+                let mut dyn_schema_field = dynamic::Field::new(
+                    field_name,
+                    type_ref.clone(),
+                    move |ctx| {
+                        let req_ctx = ctx.ctx.data::<Arc<RequestContext>>().unwrap();
+                        let field_name = &field.name;
+                        let path_name = ctx
+                            .path_node
+                            .map(|p| p.to_string())
+                            .unwrap_or(field_name.clone());
+                        match &field.resolver {
+                            None => {
                                 let ctx = EvaluationContext::new(req_ctx, &ctx);
+                                FieldFuture::from_value(
+                                    ctx.path_value(&[field_name]).map(|a| a.to_owned()),
+                                )
+                            }
+                            Some(expr) => {
+                                let expr = expr.to_owned();
+                                FieldFuture::new(
+                                async move {
+                                    let ctx = EvaluationContext::new(req_ctx, &ctx);
 
-                                let const_value = expr.eval(&ctx, &Concurrent::Sequential).await?;
+                                    let const_value =
+                                        expr.eval(&ctx, &Concurrent::Sequential).await?;
 
-                                let p = match const_value {
-                                    ConstValue::List(a) => FieldValue::list(a),
-                                    a => FieldValue::from(a),
-                                };
-                                Ok(Some(p))
-                            })
+                                    let p = match const_value {
+                                        ConstValue::List(a) => FieldValue::list(a),
+                                        a => FieldValue::from(a),
+                                    };
+                                    Ok(Some(p))
+                                }
+                                .instrument(tracing::info_span!("field::resolver", name = path_name, graphql.returnType = %type_ref)).inspect_err(|err| tracing::error!(?err)),
```

Comment:
We instrument even when telemetry is not enabled. Does it not have a performance implication? Can we conditionally instrument it?

Reply:
I moved expressions that was used only for tracing inside the macro itself. The way the macro works is wrapping those fields inside the condition block and execute it only when specific trace is enabled. So, there is still checks if tracing is enabled by fields are calculated only when enabled.

The benchmarks shows improvements after this change (without telemetry enabled):

## initial with removed tracing instrumentation

```
Running 30s test @ http://localhost:8000/graphql
  8 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.92ms    7.25ms 342.85ms   99.44%
    Req/Sec     8.03k     0.89k   19.43k    92.82%
  1929587 requests in 33.00s, 9.67GB read
  Socket errors: connect 0, read 0, write 0, timeout 96
Requests/sec:  58477.18
Transfer/sec:    300.14MB
```

## with tracing instrumentation previous

```
Running 30s test @ http://localhost:8000/graphql
  8 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.07ms    7.32ms 266.89ms   99.41%
    Req/Sec     7.47k   823.21    15.20k    91.66%
  1793795 requests in 33.63s, 8.99GB read
  Socket errors: connect 0, read 0, write 0, timeout 96
Requests/sec:  53346.59
Transfer/sec:    273.81MB
```

## after the change

```
Running 30s test @ http://localhost:8000/graphql
  8 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.97ms    8.07ms 355.39ms   99.36%
    Req/Sec     8.20k     0.89k   16.64k    90.58%
  1968957 requests in 32.98s, 9.87GB read
  Socket errors: connect 0, read 0, write 0, timeout 96
Requests/sec:  59709.01
Transfer/sec:    306.47MB
```